velox/exec/MergeJoin.cpp

```diff
@@ -261,13 +270,34 @@ void MergeJoin::addOutputRowForLeftJoin(
   ++outputSize_;
 }
 
+void MergeJoin::flattenRightProjections() {
+  auto& children = output_->children();
+
+  for (const auto& projection : rightProjections_) {
+    auto& currentVector = children[projection.outputChannel];
+    auto newFlat = BaseVector::create(
+        currentVector->type(), outputBatchSize_, operatorCtx_->pool());
+    newFlat->copy(currentVector.get(), 0, 0, outputSize_);
+    children[projection.outputChannel] = newFlat;
+  }
+  isRightFlattened_ = true;
+}
+
 void MergeJoin::addOutputRow(
     const RowVectorPtr& left,
     vector_size_t leftIndex,
     const RowVectorPtr& right,
     vector_size_t rightIndex) {
-  copyRow(left, leftIndex, output_, outputSize_, leftProjections_);
-  copyRow(right, rightIndex, output_, outputSize_, rightProjections_);
+  // All left side projections (and the right side) share the same dictionary
```

Comment:
Wondering why is this the case?

Reply:
We only create two dictionary indices buffers, one for left and one for right, and share them across left and right projections.

Reply:
Oh... I misunderstood. Somehow I though it says that we use same indices for both left and right, but here you are saying that all columns on the left use the same dictionary as well as all columns on the right use the same dictionary, but dictionaries used for left and right columns are different. Is that right?

Reply:
Exactly, will rephrase the comment to avoid confusion.